feat(agent): Create Delegated Agent installation mechanism

aroberts87 opened the pull request:
## What this PR does / why we need it:
Add the ability to install the agent in a specialized delegated manner where only a small portion of the overall nodes in the cluster communicate with the Kubernetes API server. This is accomplished by installing a modified Agent DaemonSet with an additional Agent Deployment. The Agents in the Deployment are the ones which communicate with the Kubernetes API server, but don't do much, if anything, else.

## Checklist

[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]

- [x] Title of the PR starts with type and scope, (e.g. feat(agent,node-analyzer,sysdig-deploy):)
- [x] Chart Version bumped for the respective charts
- [x] Check GithubAction checks (like lint) to avoid merge-check stoppers
- [x] All test files are added in the tests folder of their respective chart and have a "_test" suffix

Check Contribution guidelines in README.md for more insight.

aroberts87 committed:
Create Delegated Agent installation mechanism

Add the ability to install the agent in a specialized delegated
manner where only a small portion of the overall nodes in the
cluster communicate with the Kubernetes API server. This is accomplished
by installing a modified Agent DaemonSet with an additional Agent
Deployment. The Agents in the Deployment are the ones which communicate
with the Kubernetes API server, but don't do much, if anything, else.

aroberts87 committed:
Merge branch 'master' into agent_large_cluster_support

aroberts87 committed:
Merge branch 'master' into agent_large_cluster_support

# Conflicts:
#	charts/agent/Chart.yaml

aroberts87 committed:
Rename largeClusterSupport flag to delegatedAgentDeployment

aroberts87 committed:
Fix regression in prometheus.yaml file placement

* Put erroneously removed intent command back for the
  prometheus.yaml file in the Agent daemonset configmap.
* Add new unit tests to address the prometheus.yaml file
  in the agent daemonset configmap

aroberts87 committed:
Add additional label tests

aroberts87 committed:
Add unit tests for and fix various issues

Adding more unit tests unsurprisingly found issues that have now been addressed

aroberts87 committed:
Minor testcase naming update

aroberts87 committed:
Optimize delegated deployment proxy test

aroberts87 committed:
Unit test optimizations

aroberts87 committed:
Add calls to "get_if_not_in_settings" back

A regression was discovered during testing where if the collector settings were
provided in the sysdig.settings block and not via the provided parameters
(collectorSettings.collectorHost, region, etc.) then we were rendering out the
collector information for the default region (us1) as well as the content
stored in the sysdig.settings block. Since the defined parameters are rendered
first, the unintended collector information for us1 appears before the values
specified in the settings block which results in the agent attempting to
connect to the wrong collectors. This change adds the safety net back that had
covered this workflow before that had been accidentally omitted.

aroberts87 committed:
Merge branch 'master' into agent_large_cluster_support

aroberts87 committed:
Agent chart version bump

aroberts87 committed:
Fix issues with leaderelection flag

* Improperly formatted include statement has been fixed
* Per discussions with the engineering team it was decided that when
  delegatedAgentDeployment.enabled=true that the max_parallel_starts
  should really only ever be a small value like 1
* Added some unit tests around the leaderelection flags

aroberts87 committed:
Add newline to end of k8s_coldstart_test.yaml

aroberts87 committed:
Resolve issues around sysdig.disableCaptures

* sysdig_capture_enable had not been rendered in the deployment configmap
* The indentation for sysdig_capture_enabled in the daemonset configmap
  had accidentally been put one level of indentation too deep
* Added unit tests to ensure correctness going forward

aroberts87 committed:
Resovle deployment service account issue

When delegatedAgentDeployment.enabled=true was set, the Service Account for the
Agent Deployment was not set correctly, so the cointerface process was not able
to function correctly.

irozzo-1A commented on charts/agent/templates/configmap.yaml:
```diff
@@ -1,67 +1,32 @@
+    {{- include "agent.clusterName" . | indent 4 }}
+    {{- if .Values.delegatedAgentDeployment.enabled }}
+    cointerface_enabled: false
+    {{- end }}
```
nit: it could be useful to deploy the agent with cointerface enabled even with the agent deployment. It would help to have:
1. More uniform resource consumption for the agents (no delegated)
2. No need for running delegation process
If this is not the primary goal we can tackle this later

aroberts87 replied:
Hey @irozzo-1A , could you please elaborate on this? My understanding was that the whole purpose of having the deployment was to migrate cointerface functionality there. What would be purpose of having two cointerface processes running on the same node?

irozzo-1A replied:
sure @aroberts87 , the main purpose of this is to have one only Pod talking to Apiserver. This would be another use-case for average-size clusters where the Apiserver load is not a concern.
As the delegated agent consumes more resources than the others (especially memory), this would provide more uniformity over the agents controlled by the ds (it will be easier to set requests/limits). Another benefit would be not having to run the delegation election process (should be stable in the last agent versions, but we had some bugs in the past).

mavimo replied:
@irozzo-1A considering the implication (capabilities that we are going to lose) and the fact that we discourage the usage of this feature, I'll suggest to keep it only ad the solution in case of large clusters, WDYT?

irozzo-1A replied:
@mavimo I think this could turn useful for average-sized clusters, but don't consider this as a blocking point for this PR. We can address it later in case of need.

irozzo-1A commented on charts/agent/templates/configmap.yaml:
```diff
@@ -1,67 +1,32 @@
-{{- if .Values.daemonset.deploy }}
+{{- if or .Values.daemonset.deploy .Values.delegatedAgentDeployment.enabled }}
```
Shouldn't we provide the possibility to deploy the deployment without the daemonset?

aroberts87 replied:
That's an interesting point, but what would that provide the customer? Just k8s metadata?

irozzo-1A replied:
Yes, for instance, secure features that only require an agent to send global kubernetes state to BE, could benefit from this.
I'm going to double-check if there is some 😄 .

irozzo-1A commented on charts/agent/templates/configmap.yaml:
```diff
@@ -1,67 +1,32 @@
-{{- end }}
-{{- if .Values.global.sysdig.tags }}
+    {{- include "agent.clusterName" . | indent 4 }}
+    {{- if .Values.delegatedAgentDeployment.enabled }}
```
This is causing trailing spaces:
```
    new_k8s: true    
    k8s_cluster_name: agent-deployment
    cointerface_enabled: false    
```
Maybe we should do something like:
```suggestion
    {{- include "agent.clusterName" . | trim | nindent 4 }}
    {{- if .Values.delegatedAgentDeployment.enabled }}
```

irozzo-1A commented on charts/agent/templates/configmap.yaml:
```diff
@@ -1,67 +1,32 @@
+    {{- if .Values.delegatedAgentDeployment.enabled }}
+    cointerface_enabled: false
+    {{- end }}
+    {{- include "agent.connectionSettings" . | indent 4 }}
```
Same here:
```suggestion
    {{- include "agent.connectionSettings" . | trim | nindent 4 }}
```

irozzo-1A commented on charts/agent/templates/configmap-deployment.yaml:
```diff
@@ -0,0 +1,47 @@
+    feature:
+      mode: monitor_light
+    cointerface_enabled: true
+    k8s_delegated_nodes: -1
```
We could think about making this configurable

irozzo-1A commented on charts/agent/values.yaml:
```diff
@@ -275,6 +275,41 @@ tolerations:
+      - linux
+    labels: {}
+    progressDeadlineSeconds: 600
+    replicas: 2
```
I would rather use 1 as the default, to avoid having two pods to send the global kubernetes state

irozzo-1A commented on charts/agent/templates/daemonset.yaml:
```diff
@@ -1,12 +1,12 @@
-{{- if .Values.daemonset.deploy }}
+{{- if or .Values.daemonset.deploy .Values.delegatedAgentDeployment }}
```
Same here, not sure we should couple the deployment with agent daemonset

irozzo-1A reviewed (commented):
Overall LGTM, I think that the whitespaces handling in the agent config should be double-checked, as it's breaking the way it's rendered:

Got:
```
$ k get cm sysdig-agent-deployment -oyaml
apiVersion: v1
data:
  dragent.yaml: "new_k8s: true    \nk8s_cluster_name: agent-deployment    \ncollector:
    collector-staging.sysdigcloud.com\ncollector_port: 6443\nfeature:\n  mode: monitor_light\ncointerface_enabled:
    true\nk8s_delegated_nodes: -1\ntop_processes_in_sample: 0\ncontainer_engines:\n
    \ bpm: false\n  cri: false\n  docker: false\n  libvirt_lxc: false\n  lxc: false\n
    \ mesos: false\n  podman: false\n  rkt: false\napp_checks_enabled: false\njmx:\n
    \ enabled: false\nprometheus:\n  enabled: false\nstatsd:\n  enabled: false\nk8s_coldstart:\n
    \ max_parallel_cold_starts: 1\n\ncommandlines_capture:\n  enabled: false\ndrift_killer:\n
    \ enabled: false\nfalcobaseline:\n  enabled: false\nmemdump:\n  enabled: false\nnetwork_topology:\n
    \ enabled: false\nsecure_audit_streams:\n  enabled: false\nsecurity:\n  enabled:
    false\n  k8s_audit_server_enabled: true\n  k8s_audit_server_port: 7765\n  k8s_audit_server_url:
    0.0.0.0\n\n\napp_checks_enabled: false\nimds_version: 2\nlog:\n  file_priority:
    debug\n  max_size: 100\n  rotate: 25\nmetricsfile:\n  json: true\n  location:
    metrics\n"
```

Expected:
```
$ k get cm sysdig-agent -oyaml
apiVersion: v1
data:
  dragent.yaml: |
    new_k8s: true
    k8s_cluster_name: agent-deployment
    cointerface_enabled: false
    collector: collector-staging.sysdigcloud.com
    collector_port: 6443
    security:
      enabled: true
      k8s_audit_server_enabled: true
      k8s_audit_server_port: 7765
      k8s_audit_server_url: 0.0.0.0
    k8s_coldstart:
      max_parallel_cold_starts: 1


    app_checks_enabled: false
    imds_version: 2
    log:
      file_priority: debug
      max_size: 100
      rotate: 25
    metricsfile:
      json: true
      location: metrics
```

irozzo-1A committed:
Fix yaml config formatting

aroberts87 committed:
Add unit tests for Service Account sanity

aroberts87 committed:
Merge branch 'master' into agent_large_cluster_support

aroberts87 committed:
Update replica count

Based on feedback, update the replica count from 2 to 1 as this will reduce
overall resource consumption by default and not have an impact on product
performance

aroberts87 committed:
Fix captures related unit tests

Recent changes to the configmap whitespace handling inadvertently broke
some unit tests. This commit fixes those up.

aroberts87 committed:
Make k8s_delegated_nodes configurable

Add ability to override the value for k8s_delegatged_nodes when enabling the
delegated agent deployment mechanism

aroberts87 committed:
Modify version bump

The introduction of the delegated agent deployment mechanism is a large enough
change to warrant a minor version bump.

aroberts87 committed:
Merge branch 'master' into agent_large_cluster_support

# Conflicts:
#	charts/sysdig-deploy/Chart.yaml
